source/common/upstream/cluster_discovery_manager.cc

```diff
@@ -0,0 +1,171 @@
+#include "source/common/upstream/cluster_discovery_manager.h"
+
+#include <functional>
+
+#include "source/common/common/enum_to_int.h"
+
+namespace Envoy {
+namespace Upstream {
+
+namespace {
+
+using ClusterAddedCb = std::function<void(ThreadLocalCluster&)>;
+
+class ClusterCallbacks : public ClusterUpdateCallbacks {
+public:
+  ClusterCallbacks(ClusterAddedCb cb) : cb_(std::move(cb)) {}
+
+  void onClusterAddOrUpdate(ThreadLocalCluster& cluster) override { cb_(cluster); };
+
+  void onClusterRemoval(const std::string&) override {}
+
+private:
+  ClusterAddedCb cb_;
+};
+
+} // namespace
+
+ClusterDiscoveryManager::ClusterDiscoveryManager(
+    std::string thread_name, ClusterLifecycleCallbackHandler& lifecycle_callbacks_handler)
```

Comment:
Why not passing `const std::string& thread_name`?

Reply:
Because `ClusterDiscoveryManager` needs its own copy of the thread name, so I pass things by value to functions that will need the copy.

Reply:
Not really, no - it gets invoked once per thread at startup.

Reply:
The only downside I see here, is that from looking at the interface there's nothing that tells a developer that they should move the string in the call site.